proposals/2024-09-25_metadata-labels.md

```diff
@@ -0,0 +1,156 @@
+
+For example:
+
+* Querying for foo/histogram would return results that include both foo.seconds/histogram and foo.milliseconds/histogram.
+* Querying for foo.seconds would return results that include both foo.seconds/histogram and foo.seconds/counter.
+* Querying for http_server_duration would return results that include both foo.seconds/histogram and foo.milliseconds/counter.
+* Querying for an OpenTelemetry metric, such as http.server.duration, with suffixes would require querying for ”http.server.duration”.seconds/histogram. Note that suffixes are outside of quotes.
+
+This solution is not chosen because:
+
+* Requires PromQL changes (intrusive), touches on “dot” operator ideas.
+* Adding suffixes outside of quotes looks strange: {“http.server.duration”.seconds/histogram}
+* Rolling this out would be breaking for existing Prometheus users: E.g. {foo_seconds} becomes {foo.seconds/histogram}. Could this be part of OM 2.0?
+  * Mitigation: users just stay with {foo_seconds.seconds/histogram}
+* Users might be surprised by, or dislike the additional suffixes and delimiters in the metric name results
+  * Mitigation: Opt-in for query engines?
+
+### "Hide" __type__ and __unit__ labels in PromQL, instead of UI
+
+Existing UIs don't handle the `__type__` and `__unit__` labels. To mitigate this, PromQL could omit the `__type__` and `__unit__` labels from the query response. Doing this would avoid requiring UIs to update to handle the new labels.
+
+This solution is not chosen because:
+
+* It deviates from the current handling of the `__name__` label.
+* We expect metadata, like type and unit to be useful to display in the UI, and want to enable these use-cases.
+* It should be a small amount of effort to hide these labels.
+
+## Potential Future Extensions
+
+### __type__ and __unit__ from client libraries
+
+One obvious extension of this proposal would be for Prometheus clients to start sending `__type__` and `__unit__` labels with the exposition format. This would:
+
+* Allow mixing metrics with the same name, but different types and units in the same endpoint (an explicit non-goal of this proposal).
+* Allow sending unit metadata in the text exposition format, which doesn't support `# UNIT` metadata.
+    * Counter-point: It would be easier to add `# UNIT` metadata to the text exposition format.
+
+This is excluded from this proposal because:
+
+* OpenTelemetry users can use [Views](https://opentelemetry.io/docs/specs/otel/metrics/sdk/#view) to resolve collisions. That should be "good enough".
+* Doing this would require some changes to client libraries, which is significantly more work, and is harder to experiment with.
+* There can be conflicts between `# TYPE` and `# UNIT` metadata for a metric, and `__type__` and `__unit__` labels. This adds complexity to understanding the exposition format, and requires establishing rules for dealing with conflicts.
+
+### More metadata labels
```

Comment:
Unfortunately OTel resource attributes are all considered identifying today (correct me if I'm wrong David). 

There is work in progress to replace resource attributes with a new thing called `Entity`. Entity will be able to distinguish, using otel lingo, `Identifying` from `Descriptive` attributes, but this will take some years to get popular in the industry

Reply:
OTel resource attributes are all considered identifying today, but we don't treat them all as identifying when translating from OTel -> Prometheus.  Instead, we just treat service.name + service.instance.id + service.namespace as identifying.  That is correct _if_ the OTLP originates from an OTel-instrumented application, but not if it wasn't (e.g. comes from an otel collector receiver).

Anyways, what I had in mind was more things that users don't think of as "labels", like schema url, or maybe scope name + version.  I think resource attributes make sense as "real" labels, rather than metadata.